File: 2017/Day1/Day1/Day1.cpp
// Day1.cpp
// AoC 2017 Day 1: Inverse Captcha
// Author: Chi-Kit Pao
//


#include <fstream>
#include <iostream>

int main()
{
	std::fstream inFile("input.txt");
	std::string str;
	inFile >> str;

	// Part 1:
	int sum1 = 0;
	for (size_t i = 1; i < str.size(); ++i) {
		if (str[i - 1] == str[i])
			sum1 += str[i] - '0';
	}
	if (str.size() > 1 && str[0] == str[str.size() - 1])
		sum1 += str[0] - '0';

	// Part 2:
	int sum2 = 0;
	size_t halfSize = str.size() / 2;
	for (size_t i = 0; i < str.size(); ++i) {
		if (str[i] == str[(i + halfSize) % str.size()])
			sum2 += str[i] - '0';
	}

	std::cout << "Day 1: " << "\n";
	std::cout << ("Question 1: What is the solution to your captcha?\n");
	std::cout << "Answer: " << sum1 << "\n";
	std::cout << ("Question 2: What is the solution to your new captcha?\n");
	std::cout << "Answer: " << sum2 << "\n";
	std::cout << std::endl;
}

File: 2018/aoc2018/src/net/chikitpao/aoc2018/Day1.java
// Day1.java
// AoC 2018 Day 1: Chronal Calibration
// Author: Chi-Kit Pao
//

package net.chikitpao.aoc2018;

import java.io.BufferedReader;
import java.io.IOException;
import java.nio.charset.Charset;
import java.nio.file.FileSystems;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.Optional;

public class Day1 {
    public static int GetAnswer1(ArrayList<Integer> changes){
        return changes.stream().reduce(0, Integer::sum);
    }
    public static int GetAnswer2(ArrayList<Integer> changes){
        Optional<Integer> answer2 = Optional.empty();
        int frequency = 0;
        HashSet<Integer> knownFrequencies = new HashSet<>();
        knownFrequencies.add(0);
        while(true) {
            for (Integer change : changes) {
                frequency += change;
                if (answer2.isEmpty() && knownFrequencies.contains(frequency)) {
                    answer2 = Optional.of(frequency);
                    return answer2.get();
                }
                knownFrequencies.add(frequency);
            }
        }
    }
    public static void main(String[] args) {
        ArrayList<Integer> changes = new ArrayList<>();

        Charset charset = Charset.forName("US-ASCII");
        Path path = FileSystems.getDefault().getPath("input", "Day1_input.txt");
        try (BufferedReader reader = Files.newBufferedReader(path, charset)) {
            String line;
            while ((line = reader.readLine()) != null) {
                if (line.length() == 0)
                    continue;
                changes.add(Integer.parseInt(line));
            }
        } catch (IOException x) {
            System.err.format("IOException: %s", x);
        }

        System.out.println("Day 1:");
        System.out.println("Question 1: Starting with a frequency of zero, what is the resulting frequency after all "
                + "of the changes in frequency have been applied?");
        System.out.println("Answer: " + GetAnswer1(changes));
        System.out.println("Question 2: What is the first frequency your device reaches twice?");
        System.out.println("Answer: " + GetAnswer2(changes));
    }
}
